pipenv/patched/notpip/_internal/resolution/resolvelib/factory.py

```diff
@@ -295,6 +295,8 @@ def iter_index_candidate_infos() -> Iterator[IndexCandidateInfo]:
                 )
                 yield ican.version, func
 
+            raise StopIteration
```

Comment:
@frostming This one seems to need to `raise StopIteration` to prevent a bug where the `Criterion.candidates` is infinite.  What are your thoughts here?

Reply:
Oh, you are fixing the code under the vendor directory of a vendored library(`pip`). It needs to be fixed upstream.

Reply:
Actually this might not be the right approach, it bubbles as up as `RunTimError`.  Just happens to work to get out of the infinite loop, but probably isn't the right approach.

Reply:
Maybe the issue is more inside of:  https://github.com/pypa/pipenv/blob/main/pipenv/patched/notpip/_vendor/resolvelib/resolvers.py#L242-L329
We never get to the final False case for some of these examples.  It just cycles through the same back-tracking results.  It is not clear to me how to fix.

Reply:
If it is in `pip` why is the directory called `notpip`?  It seems this code is our own internal impementation of the `resolvelib` interfaces, but maybe I am misunderstanding.  That is why I asked in another thread what is `notpip`?